edit/edit.js

```diff
@@ -1476,7 +1477,28 @@ function save() {
   editor.save();
 }
 
+function isUsercssSourceCode() {
+  const sourceCode = editors[0].getValue();
+  return /==UserStyle==[\s\S]+==\/UserStyle==/i.test(sourceCode);
+}
+
 function saveSectionStyle() {
+  if (!styleId && isUsercssSourceCode() && confirm(t('styleSaveAsUsercss'))) {
+    const sourceCode = editors[0].getValue();
+    onBackgroundReady()
+      .then(() => BG.usercssHelper.save({
+        reason: 'editSave',
+        sourceCode,
+        enabled: $('#enabled').checked
+      }))
+      .then(style => {
+        window.onbeforeunload = null;
+        location.href = `/edit.html?id=${style.id}`;
+      })
+      .catch(alert);
+    return;
+  }
+
```

Comment:
I'm not sure why this isn't working... I tried pasting in [this user.css](https://gist.github.com/Mottie/82de0c2956b31d9a34c2e49ffbd4d444#file-github-selected-tab-color-uso-user-css) and I get a failed to import message. No errors in the console.

![](https://user-images.githubusercontent.com/136959/31276502-c90203e6-aa61-11e7-87ae-2ce1bc7f06be.gif)

Reply:
You can't use "Import from mozilla format" to import usercss. That feature would parse the code into multiple sections, and the error is due to those `/*[[placeholder]]*/`.

You have to paste the source code into the editor (click 'cancel') then save the style.

Reply:
I pasted directly into the editor - <kbd>Ctrl</kbd> + <kbd>v</kbd>.

Reply:
@eight04, remove that confirm box on pasting. It's terrible UX. Also, `confirm()` may be blocked by user. The *sane* approach is to check if the pasted text is a usercss in `cm.on('paste'`, for example.

Reply:
I think we can guess if the user is working on usercss by checking whether pasted text or the editor contains metadata, however, there is no way to pasted moz-format code without metadata into an empty editor.

How about this:

1. The user paste some text.
2. The text is always pasted to the editor.
3. If the text looks like moz-format, popup a toast notification to hint the user that maybe he should use "Import from mozilla format" button.
4. Or even better, provide a button in the notification which undo the paste and move the text into the importer when clicked.

The notification should only show once for a new style. If we use (4), the following paste actions can follow the previous result.

@tophf How would `confirm()` be blocked? Should I remove the `confirm()` in install-usercss.js?

Reply:
What are the most common scenarios of pasting code with `@-moz-document`? 

1. Pasting a complete standard userstyle or a section. The import dialog should automatically pop up as it does in the master branch.

2. Pasting a complete usercss or a section of it.
  2.1. When the editor is already in usercss mode: `cm.on('paste'` should do nothing so that CodeMirror handles pasting as plain text directly into the editor
  2.2. When the editor is blank in standard mode: `cm.on('paste'` checks if the pasted text looks like a usercss and proceeds with usercss routine or import dialog.

Simple and predictable. I don't understand why you suggest convoluted confirmations and hints.

>How would confirm() be blocked?

If it's displayed more than a couple of times, Chrome will add a checkbox to hide further dialogs, the user may accidentally enable it and the dialog will be blocked for good. So yeah, whenever technically possible you should avoid using js dialogs.

Reply:
> 2.2. When the editor is blank in standard mode: cm.on('paste' checks if the pasted text looks like a usercss and proceeds with usercss routine or import dialog.

But we can't determine whether the pasted text is a section of usercss or a section of standard userstyle if the text doesn't contain metadata. Should we always proceed it with import dialog? If so, I think we should mention that in the doc, maybe in a "Creating" section.

Reply:
Eh, so what? Your taking about a super rare use case. No need to complicate 99.999999% of use cases because of that.

Reply:
Also, you can check if the first section of an unsaved style in standard mode contains meta. Again, easy and simple.

Reply:
I've rethought about this, maybe we should reconsider to use a setting `newStyleFormat` to solve it. If `newStyleFormat === 'regular'`, open the multi-section editor when writing a new style. If `newStyleFormat === 'usercss'`, open the usercss editor. Pros:

1. No need to add usercss saving/pasting/parsing stuff to multi-section editor. Users who preferred regular userstyle won't notice any difference.
2. We can provide a usercss template when creating a new usercss, like Tampermonkey.
3. It can work with "Write style for..." button in the popup panel. With (2), it is possible to setup `@-moz-document` rules for the user.
	![image](https://user-images.githubusercontent.com/1324510/31314632-88b283f6-ac37-11e7-90ef-464663bbb9f7.png)

Reply:
What about another button? Add a "Write a new usercss" button next to "Write a new Style"?

And in the popup add a select:

![](https://user-images.githubusercontent.com/136959/31316472-ce4357ae-abf3-11e7-9440-d4c332b01c58.png)

Reply:
The setting may work - will it be toggled in "Advanced" section of options page?

Mottie's idea looks really awkward. I don't want to see any of usercss stuff in the main UI if I don't use it.

Reply:
I think the user won't switch the format very often, so options page should be a good place and no need to make additional button in the manager.